### PR TITLE
LSP Auto-completion

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

Suffix-based auto-completion!

* Allows searching by any valid path suffix
* Different icons for terms/types/constructors
* Automatically suffixifies selections to minimum unambiguous suffix.

Kapture 2022-10-11 at 11 22 45

## Implementation notes

* Dumps the `Names` for the current path into a name-segment based suffix-tree, such that for any _prefix_ of a valid path suffix we can list all possible names.
* Adds a handler for the LSP to provide these names to the LSP.

## Loose ends

Future improvements we could and should make:

* LSP config for selecting the number of completions to show, suffixification behaviour, etc.
* Show docs when a completion is selected (we're well set up to do this)
* Include file-local symbols and local vars in auto-completion

### CODE REVIEW

File: unison-cli/src/Unison/LSP/Completion.hs [hunk 1]
```diff
@@ -4,122 +4,236 @@
...
+  respond . maybe (Right $ InL mempty) (Right . InR) =<< runMaybeT do
+    (range, prefix) <- MaybeT $ VFS.completionPrefix (m ^. params)
+    ppe <- PPED.suffixifiedPPE <$> lift globalPPE
+    completions <- lift getCompletions
+    let defMatches = matchCompletions completions prefix
+    let (isIncomplete, defCompletions) =
+          defMatches
+            & nubOrdOn (\(p, _name, ref) -> (p, ref))
+            & fmap (over _1 Path.toText)
+            -- & ( let x = Text.dropWhileEnd (== '.') prefix
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/LSP/Completion.hs]: Note to Chris: I can just delete this now.

File: unison-cli/src/Unison/LSP/Completion.hs [hunk 2]
```diff
@@ -4,122 +4,226 @@
...
---
--- TODO:
--- * Rewrite this to use an index rather than fuzzy searching ALL names
--- * Respect ucm's current path
--- * Provide namespaces as auto-complete targets
--- * Auto-complete minimally suffixed names
--- * Include docs in completion details?
 completionHandler :: RequestMessage 'TextDocumentCompletion -> (Either ResponseError (ResponseResult 'TextDocumentCompletion) -> Lsp ()) -> Lsp ()
 completionHandler m respond =
-  respond =<< do
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/LSP/Completion.hs:32]: The old completion was incredibly naive and has been replaced entirely.

**Review by mitchellwrosen — APPROVED** [review]

File: unison-cli/src/Unison/LSP/Diagnostics.hs [hunk 3]
```diff
@@ -33,7 +33,7 @@ reportDiagnostics ::
   f Diagnostic ->
   Lsp ()
 reportDiagnostics docUri fileVersion diags = do
-  let jsonRPC = "" -- TODO: what's this for?
+  let jsonRPC = "2.0"
```
Comment (ChrisPenner) [inline unison-cli/src/Unison/LSP/Diagnostics.hs:36]: I figured out what this is for, we just literally need to set it to `"2.0"` as per the jsonRPC spec, most LSP clients don't care, but some do.

Comment (aryairani) [reply unison-cli/src/Unison/LSP/Diagnostics.hs:36]: lol

**Pull request merged**